src/execution/index/art/node.cpp

```diff
@@ -186,4 +188,208 @@ void Node::Erase(Node *&node, idx_t pos, ART &art) {
+
+		// update the prefix of r_node to only consist of the bytes after mismatch_pos
+		r_node->prefix.Reduce(mismatch_pos);
+
+		if (child_pos == DConstants::INVALID_INDEX) {
+			Node::InsertChildNode(l_node, mismatch_byte, r_node);
+			if (l_node_parent) {
+				l_node_parent->ReplaceChildPointer(l_node_pos, l_node);
+			}
+			if (r_node_parent) {
+				r_node_parent->ReplaceChildPointer(r_node_pos, nullptr);
+			}
+			r_node = nullptr;
+			return;
+		}
+
+		auto child_node = l_node->GetChild(*l_art, child_pos);
+		return Node::ResolvePrefixesAndMerge(l_art, r_art, child_node, r_node, depth + mismatch_pos, l_node, child_pos,
+		                                     r_node_parent, r_node_pos);
+	}
+
+	// prefixes differ, create new node and insert both nodes as children
+
+	// create new node
+	Node *new_node = new Node4();
+	new_node->prefix = Prefix(l_node->prefix, mismatch_pos);
+
+	// insert l_node, break up prefix of l_node
+	auto key_byte = l_node->prefix.Reduce(mismatch_pos);
+	Node4::Insert(new_node, key_byte, l_node);
+
+	// insert r_node, break up prefix of r_node
+	key_byte = r_node->prefix.Reduce(mismatch_pos);
+	Node4::Insert(new_node, key_byte, r_node);
+
+	l_node = new_node;
+	if (l_node_parent) {
+		l_node_parent->ReplaceChildPointer(l_node_pos, l_node);
+	}
+	if (r_node_parent) {
+		r_node_parent->ReplaceChildPointer(r_node_pos, nullptr);
+	}
+	r_node = nullptr;
+}
+
+void Node::Merge(ART *l_art, ART *r_art, Node *&l_node, Node *&r_node, idx_t depth, Node *&l_node_parent,
+                 idx_t l_node_pos, Node *&r_node_parent, idx_t r_node_pos) {
+
+	// always try to merge the smaller node into the bigger node
+	// because maybe there is enough free space in the bigger node to fit the smaller one
+	// without too much recursion
+
+	if (l_node->type < r_node->type) {
+		// swap subtrees to ensure that l_node has the bigger node type
+
+		auto temp_art = l_art;
+		l_art = r_art;
+		r_art = temp_art;
+
+		auto temp_node = l_node;
+		l_node = r_node;
+		r_node = temp_node;
+
+		if (l_node_parent) {
+			l_node_parent->ReplaceChildPointer(l_node_pos, l_node);
+		}
+		if (r_node_parent) {
+			r_node_parent->ReplaceChildPointer(r_node_pos, r_node);
+		}
+	}
+
+	switch (r_node->type) {
+	case NodeType::N256:
+		return Node::MergeNodeWithNode256(l_art, r_art, l_node, r_node, depth, l_node_parent, l_node_pos);
+	case NodeType::N48:
+		return Node::MergeNodeWithNode48(l_art, r_art, l_node, r_node, depth, l_node_parent, l_node_pos);
+	case NodeType::N16:
+		return Node::MergeNodeWithNode16OrNode4<Node16>(l_art, r_art, l_node, r_node, depth, l_node_parent, l_node_pos);
+	case NodeType::N4:
+		return Node::MergeNodeWithNode16OrNode4<Node4>(l_art, r_art, l_node, r_node, depth, l_node_parent, l_node_pos);
+	case NodeType::NLeaf:
+		D_ASSERT(l_node->type == NodeType::NLeaf);
+		D_ASSERT(r_node->type == NodeType::NLeaf);
+		return Leaf::Merge((l_art->IsPrimary() || l_art->IsUnique()), l_node, r_node);
```

Comment:
Add virtual at Node, and implement them at every correct Node, and use that
`virtual Node::Merge()
Node4::Merge()
...
switch(){
case NodeType::N4:
Node4::Merge()
}
```